### PR TITLE
Move Site model fields to Location model

### PR DESCRIPTION
**Pull request by HanlinMiao**
# Closes: #2954 
# What's Changed

# TODO

- [ ] Explanation of Change(s)
- [ ] Added change log fragment(s) (for more information see [the documentation](https://docs.nautobot.com/projects/core/en/stable/development/#creating-changelog-fragments))
- [ ] Attached Screenshots, Payload Example
- [ ] Unit, Integration Tests
- [ ] Documentation Updates (when adding/changing features)
- [ ] Example Plugin Updates (when adding/changing features)
- [ ] Outline Remaining Work, Constraints from Design

### CODE REVIEW

File: nautobot/dcim/filters.py [hunk 1]
```diff
@@ -333,11 +344,93 @@ class LocationFilterSet(NautobotFilterSet, StatusModelFilterSetMixin, TenancyFil
...
+    devices = NaturalKeyOrPKMultipleChoiceFilter(
+        queryset=Device.objects.all(),
+        to_field_name="name",
+        label="Devices (name or ID)",
+    )
+    has_devices = RelatedMembershipBooleanFilter(
+        field_name="devices",
+        label="Has devices",
+    )
+    # The reverse relation here is misnamed as `powerpanel`, but fixing it would be a breaking API change.
```
Comment (glennmatthews) [inline nautobot/dcim/filters.py:360]: it actually is `powerpanels` currently for Location; we should look at fixing this to `power_panels` as a part of #2520.

File: nautobot/dcim/filters.py [hunk 2]
```diff
@@ -333,11 +344,93 @@ class LocationFilterSet(NautobotFilterSet, StatusModelFilterSetMixin, TenancyFil
         field_name="location_type__content_types",
         choices=FeatureQuery("locations").get_choices,
     )
+    has_circuit_terminations = RelatedMembershipBooleanFilter(
```
Comment (glennmatthews) [inline nautobot/dcim/filters.py:347]: Should there be a `circuit_terminations` `NaturalKeyOrPKMultipleChoiceFilter` as well?

File: nautobot/dcim/filters.py [hunk 3]
```diff
@@ -333,11 +344,93 @@ class LocationFilterSet(NautobotFilterSet, StatusModelFilterSetMixin, TenancyFil
...
+        label="Has rack groups",
+    )
+    has_racks = RelatedMembershipBooleanFilter(
+        field_name="racks",
+        label="Has racks",
+    )
+    has_prefixes = RelatedMembershipBooleanFilter(
+        field_name="prefixes",
+        label="Has prefixes",
+    )
```
Comment (glennmatthews) [inline nautobot/dcim/filters.py:387]: Should there be `racks` and `prefixes` filters as well?

File: nautobot/dcim/filters.py [hunk 4]
```diff
@@ -333,11 +344,93 @@ class LocationFilterSet(NautobotFilterSet, StatusModelFilterSetMixin, TenancyFil
...
+        label="Has vlan groups",
+    )
+    has_vlans = RelatedMembershipBooleanFilter(
+        field_name="vlans",
+        label="Has vlans",
+    )
+    has_clusters = RelatedMembershipBooleanFilter(
+        field_name="clusters",
+        label="Has clusters",
+    )
```
Comment (glennmatthews) [inline nautobot/dcim/filters.py:403]: Should there be `vlans` and `clusters` filters as well?

File: nautobot/dcim/models/locations.py [hunk 5]
```diff
@@ -214,6 +216,34 @@ class Location(TreeNode, StatusModel, PrimaryModel):
...
+    facility = models.CharField(max_length=50, blank=True, help_text="Local facility ID or description")
+    asn = ASNField(
+        blank=True,
+        null=True,
+        verbose_name="ASN",
+        help_text="32-bit autonomous system number",
+    )
+    time_zone = TimeZoneField(blank=True)
+    physical_address = models.CharField(max_length=200, blank=True)
+    shipping_address = models.CharField(max_length=200, blank=True)
```
Comment (glennmatthews) [inline nautobot/dcim/models/locations.py:228]: Let's consider whether this should instead be a `TextField`? 200 chars isn't much for some addresses.

File: nautobot/dcim/models/locations.py [hunk 6]
```diff
@@ -214,6 +216,34 @@ class Location(TreeNode, StatusModel, PrimaryModel):
...
+        help_text="GPS coordinate (latitude)",
+    )
+    longitude = models.DecimalField(
+        max_digits=9,
+        decimal_places=6,
+        blank=True,
+        null=True,
+        help_text="GPS coordinate (longitude)",
+    )
+    contact_name = models.CharField(max_length=50, blank=True)
```
Comment (glennmatthews) [inline nautobot/dcim/models/locations.py:243]: Let's consider increasing this `max_length` as well.

File: nautobot/dcim/models/locations.py [hunk 7]
```diff
@@ -214,6 +216,34 @@ class Location(TreeNode, StatusModel, PrimaryModel):
...
+    contact_phone = models.CharField(max_length=20, blank=True)
```
Comment (glennmatthews) [inline nautobot/dcim/models/locations.py:244]: Let's consider increasing this `max_length` as well so as to support numbers like `+44 7624 800555 extension 12345`; would also make it so that our Location factory could use the phone number faker.

File: nautobot/dcim/tables/locations.py [hunk 8]
```diff
@@ -72,8 +72,18 @@ class Meta(BaseTable.Meta):
             "parent",
             "tenant",
             "description",
-            "tags",
             "actions",
+            "facility",
+            "asn",
+            "time_zone",
+            "physical_address",
+            "shipping_address",
```
Comment (glennmatthews) [inline nautobot/dcim/tables/locations.py:80]: Feels weird to me to let these be table columns, but if it's feature parity with what SiteTable provides now, OK.

File: nautobot/dcim/templates/dcim/location.html [hunk 9]
```diff
@@ -54,6 +74,83 @@
...
+        <div class="panel-heading">
+            <strong>Contact Info</strong>
+        </div>
+        <table class="table table-hover panel-body attr-table">
+            <tr>
+                <td>Physical Address</td>
+                <td>
+                    {% if object.physical_address %}
+                        <div class="pull-right noprint">
+                            <a href="https://maps.google.com/?q={{ object.physical_address|urlencode }}" target="_blank" class="btn btn-primary btn-xs">
```
Comment (glennmatthews) [inline nautobot/dcim/templates/dcim/location.html:87]: I'm curious, does this actually work?

File: nautobot/dcim/templates/dcim/location.html [hunk 10]
```diff
@@ -54,6 +74,83 @@
...
+            <tr>
+                <td>Shipping Address</td>
+                <td>{{ object.shipping_address|linebreaksbr|placeholder }}</td>
+            </tr>
+            <tr>
+                <td>GPS Coordinates</td>
+                <td>
+                    {% if object.latitude and object.longitude %}
+                        <div class="pull-right noprint">
+                            <a href="https://maps.google.com/?q={{ object.latitude }},{{ object.longitude }}" target="_blank" class="btn btn-primary btn-xs">
```
Comment (glennmatthews) [inline nautobot/dcim/templates/dcim/location.html:106]: I'm curious, does this actually work?

File: nautobot/dcim/tests/test_api.py [hunk 11]
```diff
@@ -416,6 +416,85 @@ def setUpTestData(cls):
...
         }
 
+    def test_time_zone_field_post_null(self):
+        """
+        Test allow_null to time_zone field on site.
+
+        See: https://github.com/nautobot/nautobot/issues/342
+        """
+        self.add_permissions("dcim.add_site")
+        url = reverse("dcim-api:site-list")
```
Comment (glennmatthews) [inline nautobot/dcim/tests/test_api.py:426]: Presumably these need to be changed to the location equivalents here and below?

File: nautobot/dcim/tests/test_filters.py [hunk 12]
```diff
@@ -179,10 +179,80 @@ def common_test_data(cls):
...
+        tenant=tenants[1],
+        description="Cube",
+        asn=65006,
+        latitude=20,
+        longitude=20,
+        comments="comment2",
+    )
+    nested_loc = Location.objects.create(
+        name="RTP South", location_type=lt1, status=status_active, parent=loc1, asn=65007
+    )
```
Comment (glennmatthews) [inline nautobot/dcim/tests/test_filters.py:248]: Can we use factories instead here?

File: nautobot/dcim/tests/test_filters.py [hunk 13]
```diff
@@ -1167,12 +1212,266 @@ def test_content_type(self):
             Location.objects.filter(location_type__content_types=ct).count(),
         )
 
+    def test_facility(self):
+        params = {"facility": ["Facility 3", "Facility 4"]}
+        self.assertEqual(self.filterset(params, self.queryset).qs.count(), 2)
```
Comment (glennmatthews) [inline nautobot/dcim/tests/test_filters.py:1217]: These should all be using `assertQuerysetEqualAndNotEmpty`.

File: nautobot/dcim/tests/test_models.py [hunk 14]
```diff
@@ -649,6 +649,21 @@ def setUp(self):
...
+            name="Site A",
+            slug="site-a",
+            status=active_status,
+            longitude=55.1234567896,
+            latitude=55.1234567896,
+        )
+        site.validated_save()
+
+        self.assertEqual(site.longitude, Decimal("55.123457"))
+        self.assertEqual(site.latitude, Decimal("55.123457"))
```
Comment (glennmatthews) [inline nautobot/dcim/tests/test_models.py:665]: Replace site with location?

File: nautobot/docs/models/dcim/location.md [hunk 15]
```diff
@@ -10,3 +10,7 @@ Once you have defined the hierarchy of Location Types that you wish to use, you
...
 
 Much like Sites, each Location must be assigned a name and operational [`status`](../../models/extras/status.md). The same default operational statuses are defined for Locations as for Sites, but as always, you can customize these to suit your needs. Locations can also be assigned to a tenant.
+
++++ 2.0.0
+
+In Nautobot v2.0, Site and Region will be collapsed in to Location Model. As a result, Location will inherent all the properties of the Site model that is going away [#2954](https://github.com/nautobot/nautobot/issues/2954). Location Model now has `asn`, `comments`, `contact_email`, `contact_name`, `contact_phone`, `facility`, `latitude`, `longitude`, `physical_address`, `shipping_address` and `time_zone` fields.
```
Comment (glennmatthews) [inline nautobot/docs/models/dcim/location.md:16]:
```suggestion
+++ 2.0.0
    Location now supports all properties previously present on the Site model, including the `asn`, `comments`, `contact_email`, `contact_name`, `contact_phone`, `facility`, `latitude`, `longitude`, `physical_address`, `shipping_address` and `time_zone` fields.
```

File: nautobot/docs/release-notes/version-2.0.md [hunk 16]
```diff
@@ -10,6 +10,10 @@ If you are a user migrating from Nautobot v1.X, please refer to the ["Upgrading
 
 ### Added
 
+#### Collapse Region and Site Models into Location ([#2517](https://github.com/nautobot/nautobot/issues/2517))
```
Comment (glennmatthews) [inline nautobot/docs/release-notes/version-2.0.md:13]: Shouldn't this reference #2954?

**Review by glennmatthews — CHANGES_REQUESTED** [review]

File: nautobot/dcim/templates/dcim/location.html (hunk 9 above)
Comment (HanlinMiao) [reply nautobot/dcim/templates/dcim/location.html:87]: It does here

File: nautobot/dcim/templates/dcim/location.html (hunk 10 above)
Comment (HanlinMiao) [reply nautobot/dcim/templates/dcim/location.html:106]: It does here as well

File: nautobot/dcim/filters.py (hunk 2 above)
Comment (HanlinMiao) [reply nautobot/dcim/filters.py:347]: Those are in Meta.fields just like Site model now, and they seemed to be mapped to ModelMultipleChoiceFilter.

Comment (HanlinMiao) [reply nautobot/dcim/filters.py:347]: So I added `racks`, `clusters`, and `vlans` filters as `NaturalKeyOrPKMultipleChoiceFilter`. I am not too sure the other field for `NaturalKeyOrPKMultipleChoiceFilter` to map to for `circuit_terminations` and `prefixes`. Both models do not have slug or a sensible "NaturalKey" to map to.

Comment (glennmatthews) [reply nautobot/dcim/filters.py]: I think that makes sense for now, yeah. Thanks. :-)

File: nautobot/dcim/tests/test_filters.py [hunk 17]
```diff
@@ -185,69 +185,49 @@ def common_test_data(cls):
...
+    loc4 = Location.objects.filter(location_type=lt4).first()
+    loc4.parent = loc3
+    loc4.tenant = tenants[1]
+    loc4.description = "Cube"
+    loc4.asn = 65006
+    loc4.latitude = 20
+    loc4.longitude = 20
+    loc4.comments = "comment2"
+    nested_loc = Location.objects.filter(location_type__nestable=True, parent=loc1).first()
+    nested_loc.asn = 65007
```
Comment (glennmatthews) [inline nautobot/dcim/tests/test_filters.py:222]: Can we instead update `LocationFactory` so that it automatically populates these new fields appropriately in all cases? I'm okay if you want to do that as a separate PR, but it's definitely where we want to end up -- should be able to copy a lot of it from `SiteFactory`...

File: nautobot/docs/release-notes/version-2.0.md [hunk 18]
```diff
@@ -10,6 +10,12 @@ If you are a user migrating from Nautobot v1.X, please refer to the ["Upgrading
...
+
+##### Move Site Fields to Location ([#2954](https://github.com/nautobot/nautobot/issues/2954))
```
Comment (glennmatthews) [inline nautobot/docs/release-notes/version-2.0.md:15]:
```suggestion
#### Added Site Fields to Location ([#2954](https://github.com/nautobot/nautobot/issues/2954))
```

#2517 will probably be a `Changed` or `Removed` item rather than `Added`.

File: nautobot/dcim/tests/test_views.py [hunk 19]
```diff
@@ -317,6 +317,18 @@ def setUpTestData(cls):
...
             "status": active.pk,
             "tenant": tenant.pk,
+            "facility": "Facility X",
+            "asn": 65001,
+            "time_zone": pytz.UTC,
+            "physical_address": "742 Evergreen Terrace, Springfield, USA",
+            "shipping_address": "742 Evergreen Terrace, Springfield, USA",
+            "latitude": Decimal("35.780000"),
+            "longitude": Decimal("-78.642000"),
+            "contact_name": "Hank Hill",
```
Comment (glennmatthews) [inline nautobot/dcim/tests/test_views.py:327]: Mixing Simpsons with King of the Hill... 😿

File: nautobot/docs/models/dcim/location.md [hunk 20]
```diff
@@ -10,3 +10,7 @@ Once you have defined the hierarchy of Location Types that you wish to use, you
     At present, Locations fill the conceptual space between the more abstract Region and Site models and the more concrete Rack Group model. In a future Nautobot release, some or all of these other models may be collapsed into Locations. That is to say, in the future you might not deal with Regions and Sites as distinct models, but instead your Location Type hierarchy might include these higher-level categories, becoming something like Country ← City ← Site ← Building ← Floor ← Room.
 
 Much like Sites, each Location must be assigned a name and operational [`status`](../../models/extras/status.md). The same default operational statuses are defined for Locations as for Sites, but as always, you can customize these to suit your needs. Locations can also be assigned to a tenant.
+
++++ 2.0.0
+
+Location now supports all properties previously present on the Site model, including the `asn`, `comments`, `contact_email` `contact_name`, `contact_phone`, `facility`, `latitude`, `longitude`, `physical_address`, `shipping_address` and `time_zone` fields.
```
[the remark this answers is not in the document]
Comment (HanlinMiao) [reply nautobot/docs/models/dcim/location.md:16]: Could you give an example for what a callout box might look like?

[the remark this answers is not in the document]
Comment (glennmatthews) [reply nautobot/docs/models/dcim/location.md:16]: Changing it like the above *should* render correctly - see https://raw.githubusercontent.com/nautobot/nautobot/develop/nautobot/docs/models/dcim/locationtype.md for an example callout, and compare to the rendered `Added in version 1.5.0` callout at https://docs.nautobot.com/projects/core/en/stable/core-functionality/sites-and-racks/#location-types

File: nautobot/dcim/api/serializers.py [hunk 21]
```diff
@@ -303,6 +303,13 @@ class LocationSerializer(NautobotModelSerializer, TaggedModelSerializerMixin, St
     tenant = NestedTenantSerializer(required=False, allow_null=True)
     site = NestedSiteSerializer(required=False, allow_null=True)
     tree_depth = serializers.SerializerMethodField(read_only=True)
+    time_zone = TimeZoneSerializerField(required=False, allow_null=True)
+    circuit_count = serializers.IntegerField(read_only=True)
+    device_count = serializers.IntegerField(read_only=True)
+    prefix_count = serializers.IntegerField(read_only=True)
+    rack_count = serializers.IntegerField(read_only=True)
+    virtualmachine_count = serializers.IntegerField(read_only=True)
+    vlan_count = serializers.IntegerField(read_only=True)
```
Comment (glennmatthews) [inline nautobot/dcim/api/serializers.py:312]: Since these aren't model fields, you probably need to update `LocationViewSet.queryset` to include the appropriate `annotate(count_related(...), ...)` calls, similar to what's on `SiteViewSet.queryset` today.

**Review by glennmatthews — APPROVED** [review]
